### PR TITLE
feat(OnyxButton, OnyxCheckbox, OnyxRadioButton, OnyxSwitch): Implement loading state

### PR DESCRIPTION
**Pull request by MajaZarkova**
relates to #393 

Implement loading state

## Checklist

- [x] The added / edited code has been documented with [JSDoc](https://jsdoc.app/about-getting-started)
- [x] All changes are documented in the documentation app (folder `apps/docs`)
- [x] If a new component is added, at least one [Playwright screenshot test](https://github.com/SchwarzIT/onyx/actions/workflows/playwright-screenshots.yml) is added
- [x] A changeset is added with `npx changeset add` if your changes should be released as npm package (because they affect the library usage)

### CODE REVIEW

File: .changeset/smart-pillows-wave.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"sit-onyx": minor
+---
+
+Implement loading state
```
Comment (larsrickert) [inline .changeset/smart-pillows-wave.md:5]:
```suggestion
Add property `loading` for OnyxButton, OnyxCheckbox, OnyxRadioButton and OnyxSwitch
```

We should be as detailed as possible for the changelogs so users know what happened :)

File: packages/sit-onyx/src/components/OnyxButton/OnyxButton.vue [hunk 2]
```diff
@@ -23,11 +25,12 @@ const emit = defineEmits<{
...
     :class="[`onyx-button--${props.variation}`, `onyx-button--${props.mode}`]"
-    :disabled="props.disabled"
+    :disabled="props.disabled || props.loading"
     @click="emit('click')"
   >
-    <OnyxIcon v-if="props.icon" :icon="props.icon" size="24px" />
-    <span class="onyx-button__label onyx-truncation-ellipsis">{{ props.label }}</span>
+    <OnyxIcon v-if="props.icon && !props.loading" :icon="props.icon" size="24px" />
+    <OnyxLoadingIndicator v-if="props.loading" class="onyx-button__loading" type="dots" />
+    <span v-else class="onyx-button__label onyx-truncation-ellipsis">{{ props.label }}</span>
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxButton/OnyxButton.vue:33]:
```suggestion
    <OnyxIcon v-if="props.icon && !props.loading" :icon="props.icon" />
    <OnyxLoadingIndicator v-if="props.loading" class="onyx-button__loading" />
    <span v-else class="onyx-button__label onyx-truncation-ellipsis">{{ props.label }}</span>
```

We don't need to set the props if they are the default

File: packages/sit-onyx/src/components/OnyxButton/OnyxButton.vue [hunk 3]
```diff
@@ -57,7 +60,7 @@ $button-height: 2.5rem;
   color: var(--onyx-button-text-color);
 
   &--primary {
-    &:disabled {
+    &:disabled &:not(.onyx-button__loading) {
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxButton/OnyxButton.vue:63]: disabled state no longer works when both loading and disabled is set

File: packages/sit-onyx/src/components/OnyxButton/types.ts [hunk 4]
```diff
@@ -7,6 +7,10 @@ export type OnyxButtonProps = {
    * If the button should be disabled or not.
    */
   disabled?: boolean;
+  /**
+   * Shows a loading indicator.
+   */
+  loading?: boolean;
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxButton/types.ts:13]: Should we add a story for the Storybook docs?

File: packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx [hunk 5]
```diff
@@ -98,6 +98,17 @@ test("should render required", async ({ mount, makeAxeBuilder }) => {
   expect(accessibilityScanResults.violations).toEqual([]);
 });
 
+test("should render loading", async ({ mount, makeAxeBuilder }) => {
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx:101]: I think this is already covered by the screenshot test

File: packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx [hunk 6]
```diff
@@ -263,14 +274,15 @@ test(
...
+          loading={state === "loading"}
           hideLabel={labeled === "unlabeled"}
         />,
         { useOptional: state === "optional" },
       );
 
       const checkbox = component.getByRole("checkbox");
       if (focusState === "focus-visible") await page.keyboard.press("Tab");
-      if (focusState === "hover") await checkbox.hover();
+      if (focusState === "hover" && state != "loading") await checkbox.hover();
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx:285]:
```suggestion
      if (focusState === "hover") await checkbox.hover();
```

We should still test hover to guarantee that it looks correct

File: packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.vue [hunk 7]
```diff
@@ -195,6 +203,12 @@ $input-size: 1rem;
     font-size: 1rem;
     line-height: 1.5rem;
   }
+
+  &__loading {
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.vue:207]: text color is not set correctly when loading + disabled

File: packages/sit-onyx/src/components/OnyxCheckboxGroup/OnyxCheckboxGroup.stories.ts [hunk 8]
```diff
@@ -30,6 +30,7 @@ export const Default = {
       { label: "Initially checked", id: "id-2" },
       { label: "Required", id: "id-3", required: true },
       { label: "Disabled", id: "id-4", disabled: true },
+      { label: "Loading", id: "id-4", loading: true },
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxCheckboxGroup/OnyxCheckboxGroup.stories.ts:33]:
```suggestion
      { label: "Loading", id: "id-5", loading: true },
```

File: packages/sit-onyx/src/components/OnyxRadioButton/OnyxRadioButton.ct.tsx [hunk 9]
```diff
@@ -146,7 +147,7 @@ test(
 
       const radioInput = component.getByRole("radio");
       if (focusState === "focus-visible") await page.keyboard.press("Tab");
-      if (focusState === "hover") await radioInput.hover();
+      if (focusState === "hover" && state != "loading") await radioInput.hover();
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxRadioButton/OnyxRadioButton.ct.tsx:150]: Same here

File: packages/sit-onyx/src/components/OnyxRadioButton/OnyxRadioButton.vue [hunk 10]
```diff
@@ -146,6 +154,14 @@ $input-size: var(--onyx-spacing-md);
       border-radius: 100%;
     }
   }
+
+  &__loading {
+    color: var(--onyx-color-text-icons-primary-intense);
+    width: 1rem;
+    height: 1rem;
+    margin: var(--onyx-spacing-sm);
+    align-self: center;
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxRadioButton/OnyxRadioButton.vue:163]: Why is this needed?

File: packages/sit-onyx/src/components/OnyxSwitch/types.ts [hunk 11]
```diff
@@ -13,6 +13,10 @@ export type OnyxSwitchProps = {
    * Whether to disable the switch and prevent user interaction.
    */
   disabled?: boolean;
+  /**
+   * Shows a loading indicator.
+   */
+  loading?: boolean;
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxSwitch/types.ts:19]: Should we also add a story for the Storybook docs?

**Review by larsrickert — CHANGES_REQUESTED** [review]

File: packages/sit-onyx/src/components/OnyxSwitch/OnyxSwitch.vue [hunk 12]
```diff
@@ -75,12 +77,17 @@ const requiredMarkerClass = computed(() => {
...
       :aria-label="props.hideLabel ? props.label : undefined"
-      :disabled="props.disabled"
+      :disabled="props.disabled || props.loading"
       :required="props.required"
     />
     <span class="onyx-switch__container">
       <span class="onyx-switch__icon">
-        <OnyxIcon :icon="isChecked ? checkSmall : xSmall" size="24px" />
+        <OnyxLoadingIndicator
+          v-if="!props.disabled && props.loading"
```
[the remark this answers is not in the document]
Comment (JoCa96) [reply packages/sit-onyx/src/components/OnyxSwitch/OnyxSwitch.vue:86]: Yes, just make sure that the switch is always disabled when it is loading:

```suggestion
          v-if="props.loading"
```

File: packages/sit-onyx/src/components/OnyxRadioButton/OnyxRadioButton.vue (hunk 10 above)
Comment (MajaZarkova) [reply packages/sit-onyx/src/components/OnyxRadioButton/OnyxRadioButton.vue:163]: This is how it looks like without it

image

File: packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx (hunk 6 above)
Comment (MajaZarkova) [reply packages/sit-onyx/src/components/OnyxCheckbox/OnyxCheckbox.ct.tsx]: But the tests are failing if I remove that. Also all the screenshot tests are failing for my branch 😞

File: packages/sit-onyx/src/utils/playwright.tsx [hunk 13]
```diff
@@ -109,7 +109,7 @@ export const executeScreenshotsForAllStates = <
   // therefore, we will split the permutations to only run a maximum of X permutations inside a single test
   const batches: (typeof permutations)[] = [];
   while (permutations.length > 0) {
-    batches.push(permutations.splice(0, 100));
+    batches.push(permutations.splice(0, 50));
```
Comment (larsrickert) [inline packages/sit-onyx/src/utils/playwright.tsx:112]:
```suggestion
    batches.push(permutations.splice(0, 100));
```

Lets change this back, I did this change when I tried to fix the pipeline but since it had no effect we can revert it :)

File: packages/sit-onyx/src/components/OnyxSwitch/OnyxSwitch.vue [hunk 14]
```diff
@@ -184,10 +191,16 @@ $input-width: calc(2 * $icon-size - 2 * $container-padding);
...
       color: var(--onyx-color-text-icons-neutral-soft);
+      height: $icon-size;
+      width: $icon-size;
 
       .onyx-icon {
         --icon-size: #{$icon-size};
       }
+
+      .onyx-switch__spinner {
+        color: var(--onyx-color-text-icons-neutral-soft);
```
Comment (larsrickert) [inline packages/sit-onyx/src/components/OnyxSwitch/OnyxSwitch.vue:202]:
```suggestion
        padding: var(--onyx-1px-in-rem);
```

The loading indicator inherits the font color so since we already set it in "onyx-switch__icon" we don't need to define it here. But imho we should add a small padding so the indicator is better visible :)

**Review by larsrickert — CHANGES_REQUESTED** [review]

File: packages/sit-onyx/src/components/OnyxSwitch/OnyxSwitch.vue (hunk 14 above)
Comment (larsrickert) [reply packages/sit-onyx/src/components/OnyxSwitch/OnyxSwitch.vue:202]: In figma it also has a padding/gap :)

**Review by larsrickert — APPROVED** [review]
